### ISSUE TITLE
t.web2.client.http.HTTPClientChannelRequest.connectionLost does nothing

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2304]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)| @ghazel reported|
|-|-|
|Trac ID|trac#2304|
|Type|defect|
|Created|2006-12-18 20:25:06Z|

The comment says it all:

```
    def connectionLost(self, reason):
        ### FIXME!
        pass
```

I believe it should be:

```
    def connectionLost(self, reason):
        self._error(reason)
```

<details><summary>Searchable metadata</summary>

```
trac-id__2304 2304
type__defect defect
reporter__ghazel ghazel
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1166473506000000 1166473506000000
changetime__1192781817000000 1192781817000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="scunningham's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|scunningham commented|
|-|-|

There are other issues with HTTPClientChannelRequest.  For example, if the response headers have already been received, the object will create a stream object, hook it up, and send the stream object to the owner via callback on self.responseDefer.  However, if an error occurs during the streaming process (ie. connection drops or theres a chunk parsing problem), that error cannot be propagated to the owner because self.responseDefer has already been called.

Probably need to do something like this:

old__error = http.HTTPClientChannelRequest._error

def _error(self, err):

    self.abortParse()

    if self.stream:

        assert(self.responseDefer.called)

        self.stream.finish(failure=err)

    else:

        self.responseDefer.errback(err)

http.HTTPClientChannelRequest._error = _error


old_abortWithError = http.HTTPClientChannelRequest._abortWithError

def _abortWithError(self, errcode, text):

    self._error(http.ProtocolError(text))

http.HTTPClientChannelRequest._abortWithError = _abortWithError

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [21315]) Fix connectionLost and add a simple test.

Refs [#2304](https://github.com/twisted/twisted/issues/2304)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [21316]) Improve cosmetics

Refs [#2304](https://github.com/twisted/twisted/issues/2304)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

I fix the main ticket problem in web2-client-connectionlost-2304, in revision [21315]. The second revision is here for aesthetic concerns.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_5"></a>|@dreid set owner to @therve|
|-|-|

Can we also address scunningham's issue with _error not doing the right thing when we've already fired responseDefer and started writing to the stream?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

(In [21331]) Correct the behavior for the stream finish.

Refs [#2304](https://github.com/twisted/twisted/issues/2304)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @jyknight|
|-|-|

OK that was not that hard, the formating of the comment just scared me away :).

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @dreid<br>@dreid set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_9"></a>|@dreid set owner to @therve<br>@dreid set status to `new`|
|-|-|

First off, thank you thank you thank you thank you for giving twisted.web2.client some much needed love.

All the work you've done so far looks very good.  However I'm noticing that _abortWithError is pretty useless on the HTTP client end.  It's used extensively in the  HTTPParser super class, however for the client it's implementation should just be replaced with a call to self._error(ProtocolError(text))

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

(In [21348]) Change body of _abortError

Refs [#2304](https://github.com/twisted/twisted/issues/2304)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @dreid|
|-|-|

Thanks for the review, the modification is made above.

Back for the last round ? :)

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_12"></a>|@dreid set owner to @therve|
|-|-|

I vote merge it.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set status to `closed`|
|-|-|

(In [21366]) Merge web2-client-connectionlost-2304

Author: therve
Reviewer: dreid
Fixes [#2304](https://github.com/twisted/twisted/issues/2304)

Fix the connectionLost method of web2 client request so that errors are
propagated to the response deferred. Also it handles the case where the headers
are sent and the error happens during the streaming (after the response
deferred).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
